nemo/collections/asr/parts/utils/confidence_metrics.py

```diff
@@ -13,102 +13,246 @@
     """Compute Normalized Cross Entropy (NCE) from prediction scores. Also known as the Normalized Mutual Information.
 
     NCE measures how close the correct prediction scores are to one and the incorrect prediction scores are to zero.
     Negative NCE values indicate that the classifier performs worse than the setting all prediction scores
     as the proportion of correct predictions.
+
+    Note: If only one class is present in y_true, 0.5 is returned.
     """
-    p = sum(y_true) / len(y_true)
+    y_true = np.array(y_true)
+    y_score = np.array(y_score)
+    assert len(y_true) == len(y_score)
+    assert np.all(y_true >= 0) and np.all(y_true <= 1)
+    if np.all(y_true == 0) or np.all(y_true == 1):
+        return -math.inf
+    p = y_true.mean()
     eps = 1e-15
     Hp = -(math.log(p + eps) * p + math.log(1 - p + eps) * (1 - p))
     return (Hp - log_loss(y_true, y_score)) / Hp
 
 
-def ece(y_true, y_score, n_bins=100):
+def ece(
+    y_true: Union[List[int], np.ndarray],
+    y_score: Union[List[float], np.ndarray],
+    n_bins: int = 100,
+    return_curve: bool = False,
+) -> Union[float, Tuple[float, Tuple[List[int], List[float]]]]:
     """Compute Expected Calibration Error (ECE) from prediction scores.
 
     ECE measures how close the correct prediction scores are to one and the incorrect prediction scores are to zero.
     ECE ranges from zero to one with the best value zero (the lower the value, the better).
     """
     y_true = np.array(y_true)
     y_score = np.array(y_score)
+    assert len(y_true) == len(y_score)
+    assert np.all(y_true >= 0) and np.all(y_true <= 1)
     py = np.array([1 - y_score, y_score]).T
     acc, conf = np.zeros(n_bins), np.zeros(n_bins)
     Bm = np.zeros(n_bins)
+    ece_curve = []
+    thresholds = []
     for m in range(n_bins):
         a, b = m / n_bins, (m + 1) / n_bins
         threshold = (a + b) / 2
+        thresholds.append(threshold)
         py_index = (py.T[1] >= threshold).astype(int)
         py_value = py[np.arange(len(py_index)), py_index]
         bin_range = ((py_value > a) & (py_value <= b)).nonzero()[0]
         Bm[m] = len(bin_range)
         if Bm[m] > 0:
-            acc[m] = (py_index[bin_range] == y_true[bin_range]).sum()
-            conf[m] = py_value[bin_range].sum()
-        if Bm[m] != 0:
-            acc[m] /= Bm[m]
-            conf[m] /= Bm[m]
-    ece = 0
-    for m in range(n_bins):
-        ece += Bm[m] * np.abs((acc[m] - conf[m]))
-    return ece / sum(Bm)
+            acc[m] = (py_index[bin_range] == y_true[bin_range]).sum() / Bm[m]
+            conf[m] = py_value[bin_range].sum() / Bm[m]
+        ece_curve.append(Bm[m] * np.abs(acc[m] - conf[m]))
+    ece = sum(ece_curve) / sum(Bm)
+    if return_curve:
+        return ece, (thresholds, ece_curve)
+    else:
+        return ece
 
 
-def auc_yc(y_true, y_score, return_std_maximum=False, return_curve=False, n_bins=100):
+def auc_yc(
```

Comment:
Add tests for all metrics

Reply:
Is [tests/collections/asr/confidence/test_asr_confidence_metrics.py](https://github.com/NVIDIA/NeMo/pull/6810/files#diff-13a236310d7d06a28eda4d6b741249f5dd358b7976a1b3a3ef5280b1355174d1) not enough?